autoemulate/emulators/neural_networks/rbf.py

```diff
@@ -0,0 +1,331 @@
+        r = self.norm_function(diff)
+
+        # Apply parameter, eps_r has size B x num_kernels
+        eps_r = self.log_shapes.exp().expand(batch_size, self.num_kernels) * r
+
+        # Apply radial basis function; rbf has size B x num_kernels
+        rbfs = self.radial_function(eps_r)
+
+        # Apply normalization
+        # (check https://en.wikipedia.org/wiki/Radial_basis_function_network)
+        if self.normalization:
+            # 1e-9 prevents division by 0
+            rbfs = rbfs / (1e-9 + rbfs.sum(dim=-1)).unsqueeze(-1)
+
+        # Take linear combination
+        out = self.weights.expand(
+            batch_size, self.out_features_dim, self.num_kernels
+        ) * rbfs.view(batch_size, 1, self.num_kernels)
+
+        return out.sum(dim=-1)
+
+    @property
+    def get_kernels_centers(self):
+        """Returns the centers of the kernels"""
+        return self.kernels_centers.detach()
+
+    @property
+    def get_weights(self):
+        """Returns the linear combination weights"""
+        return self.weights.detach()
+
+    @property
+    def get_shapes(self):
+        """Returns the shape parameters"""
+        return self.log_shapes.detach().exp()
+
+
+class RBFModule(TorchModule):
+    """
+    Radial Basis Function Layer module for NeuralNetRegressor
+
+    """
+
+    def __init__(
+        self,
+        input_size: int = None,
+        output_size: int = None,
+        random_state: int = None,
+        num_kernels: int = 1,
+        radial_function: Callable[[torch.Tensor], torch.Tensor] = rbf_gaussian,
+        norm_function: Callable[[torch.Tensor], torch.Tensor] = l_norm,
+        normalization: bool = True,
+    ):
+        super(RBFModule, self).__init__(
+            module_name="rbf",
+            input_size=input_size,
+            output_size=output_size,
+            random_state=random_state,
+        )
+        assert num_kernels >= 1
+        modules = [
+            RBFLayer(
+                in_features_dim=self.input_size,
+                num_kernels=num_kernels,
+                out_features_dim=output_size,
+                radial_function=radial_function,
+                norm_function=norm_function,
+                normalization=normalization,
+            )
+        ]
+        self.model = nn.Sequential(*modules)
+
+    def get_grid_params(self, search_type: str = "random"):
+        param_space = {
+            "max_epochs": np.arange(10, 110, 10).tolist(),
+            "batch_size": np.arange(2, 128, 2).tolist(),
+            "module__num_kernels": np.arange(1, 11).tolist(),
+            "module__radial_function": [
+                rbf_gaussian,
+                rbf_linear,
+                rbf_multiquadric,
+                rbf_inverse_quadratic,
+                rbf_inverse_multiquadric,
+            ],
+            "optimizer": [torch.optim.AdamW, torch.optim.SGD],
+            "optimizer__weight_decay": (1 / 10 ** np.arange(1, 9)).tolist(),
+        }
```

Comment:
looks good, but doesn't work great yet. I don't have a good intuition whether radial basis function networks will work well at all for our datasets, but maybe we can try to optimise this in the next sprint.